src/audio/igo_nr/igo_nr.c

```diff
@@ -0,0 +1,804 @@
+
+	err = igo_nr_verify_params(dev, pcm_params);
+	if (err < 0) {
+		comp_err(dev, "igo_nr_params(): pcm params verification failed.");
+		return -EINVAL;
+	}
+
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
+				  sink_list);
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer,
+				source_list);
+
+	/* set source/sink_frames/rate */
+	cd->source_rate = sourceb->stream.rate;
+	cd->sink_rate = sinkb->stream.rate;
+	if (!cd->sink_rate) {
+		comp_err(dev, "igo_nr_params(), zero sink rate");
+		return -EINVAL;
+	}
+
+	cd->sink_frames = dev->frames;
+	cd->source_frames = ceil_divide(dev->frames * cd->source_rate,
+					cd->sink_rate);
+
+	/* Use empirical add +10 for target frames number to avoid xruns and
+	 * distorted audio in beginning of streaming. It's slightly more than
+	 * min. needed and does not increase much peak load and buffer memory
+	 * consumption. The copy() function will find the limits and process
+	 * less frames after the buffers levels stabilize.
+	 */
+	cd->source_frames_max = cd->source_frames + 10;
+	cd->sink_frames_max = cd->sink_frames + 10;
+	cd->frames = MAX(cd->source_frames_max, cd->sink_frames_max);
+
+	comp_info(dev, "igo_nr_params(), source_rate=%u, sink_rate=%u, source_frames_max=%d, sink_frames_max=%d",
+		  cd->source_rate, cd->sink_rate,
+		  cd->source_frames_max, cd->sink_frames_max);
+
+	/* The igo_nr supports sample rate 16000 only. */
+	switch (sourceb->stream.rate) {
+	case 16000:
+		comp_info(dev, "igo_nr_params(), sample rate = 16000");
+		cd->invalid_param = false;
+		break;
+	default:
+		comp_err(dev, "igo_nr_params(), invalid sample rate");
+		cd->invalid_param = true;
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static inline void igo_nr_set_chan_process(struct comp_dev *dev, int32_t chan)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	if (!cd->process_enable[chan])
+		cd->process_enable[chan] = true;
+}
+
+static inline void igo_nr_set_chan_passthrough(struct comp_dev *dev, int32_t chan)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	if (cd->process_enable[chan]) {
+		cd->process_enable[chan] = false;
+		IgoLibInit(cd->p_handle, &cd->igo_lib_config, &cd->config.igo_params);
+	}
+}
+
+static int32_t igo_nr_cmd_get_data(struct comp_dev *dev,
+				   struct sof_ipc_ctrl_data *cdata,
+				   int32_t max_size)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int32_t ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "igo_nr_cmd_get_data(), SOF_CTRL_CMD_BINARY");
+		ret = comp_data_blob_get_cmd(cd->model_handler, cdata, max_size);
+		break;
+	default:
+		comp_err(dev, "igo_nr_cmd_get_data() error: invalid cdata->cmd", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int32_t igo_nr_cmd_get_value(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int32_t j;
+	int32_t ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_ENUM:
```

Comment:
just curious why you have both enum and switch? is one broken?

Reply:
Nope. I just added both of them during the topology debugging. No harm to keep them there in case we might need one of them in the future.

Reply:
Lets remove the enum since ALSA will map the switch as an ON/OFF type control. @fuyuntsuo this can be done as an update.

Reply:
This issue has been resolved in #4152.
@lgirdwood I don't quite follow. SWITCH is the ON/OFF type control but ENUM is the multiple choice type control. What's the reason to keep SWITCH only? Will ENUM be deprecated in the future?